s3/s3_operations.cpp

```diff
@@ -245,19 +245,44 @@ namespace irods_s3 {
             }
         }
 
-
-        // if this is a replication and we're the destination, get the data size from the source dataObjInfo
+        // special treatment for replication - get data_size from source and
+        // number_of_threads from destination
         if (oprType == REPLICATE_DEST) {
+            bool found_data_size = false;
+            bool found_number_of_threads = false;
 
             for (int i = 0; i < NUM_L1_DESC; ++i) {
-
-                if (L1desc[i].inuseFlag && L1desc[i].dataObjInp && L1desc[i].dataObjInfo
-                       && L1desc[i].dataObjInp->objPath == file_obj->logical_path()
-                       && L1desc[i].dataObjInp->oprType == REPLICATE_SRC ) {
-
+                // get the data size from source dataObjInfo
+                if (!found_data_size && L1desc[i].inuseFlag && L1desc[i].dataObjInp && L1desc[i].dataObjInfo &&
+                    L1desc[i].dataObjInp->objPath == file_obj->logical_path() &&
+                    L1desc[i].dataObjInp->oprType == REPLICATE_SRC)
+                {
                     data_size = L1desc[i].dataObjInfo->dataSize;
                     rodsLog(developer_messages_log_level, "%s:%d (%s) [[%lu]] repl to s3 destination.  setting data_size to %zd\n",
                             __FILE__, __LINE__, __FUNCTION__, thread_id, data_size);
+
+                    found_data_size = true;
+                }
+
+                // get the number_of_threads from destination dataObjInp
+                if (!found_number_of_threads && L1desc[i].inuseFlag && L1desc[i].dataObjInp && L1desc[i].dataObjInfo &&
+                    L1desc[i].dataObjInp->objPath == file_obj->logical_path() &&
+                    L1desc[i].dataObjInp->oprType == REPLICATE_DEST)
+                {
+                    number_of_threads = L1desc[i].dataObjInp->numThreads;
+                    rodsLog(developer_messages_log_level,
+                            "%s:%d (%s) [[%lu]] repl to s3 destination.  setting number_of_threads to %d",
+                            __FILE__,
+                            __LINE__,
+                            __FUNCTION__,
+                            thread_id,
+                            number_of_threads);
+
+                    found_number_of_threads = true;
+                }
+
+                // once we have both pieces of information break out of for loop
+                if (found_data_size && found_number_of_threads) {
```

Comment:
Are these allowed to be set on different iterations of the loop?
Should they be tied to the same L1desc?

Reply:
No, because one bit of information is taken from the replication source and the other from the replication destination.

Reply:
I think the loop allows for that possibility.